frontend/catalyst/utils/jax_extras.py

```diff
@@ -463,10 +474,23 @@ def abstractify(args, kwargs):
         in_type = infer_lambda_input_type(axes_specs, flat_args)
         return in_type, in_tree
 
+    # TODO: See the `_gather_shape_rule_dynamic` comment. Remove once the upstream change is
+    # applied.
+    gather2_p = standard_primitive(
+        _gather_shape_rule_dynamic,
+        _gather_dtype_rule,
+        "gather",
+        weak_type_rule=_argnum_weak_type(0),
+    )
+    register_lowering(gather2_p, _gather_lower)
+
     @wraps(fun)
     def make_jaxpr_f(*args, **kwargs):
         # TODO: re-use `deduce_avals` here.
-        with Patcher((jax._src.interpreters.partial_eval, "get_aval", get_aval2)), ExitStack():
+        with Patcher(
+            (jax._src.interpreters.partial_eval, "get_aval", get_aval2),
+            (jax._src.lax.slicing, "gather_p", gather2_p),
```

Comment:
@dime10 @grwlf here we have an issue, the new rule does not define jvp and therefore it is not compatible with grad or jacobian transformations ad.defjvp(gather_p, _gather_jvp_rule, None) see the original gather_p

```
gather_p = standard_primitive(
    _gather_shape_rule, _gather_dtype_rule, 'gather',
    weak_type_rule=_argnum_weak_type(0))
ad.defjvp(gather_p, _gather_jvp_rule, None)
ad.primitive_transposes[gather_p] = _gather_transpose_rule
batching.primitive_batchers[gather_p] = _gather_batching_rule
pe.padding_rules[gather_p] = _gather_pad_rule
```

Reply:
@rmoyard does this result in a user-facing bug?

Reply:
Yes you cannot use jax.grad inside qjit when a gather operation is created, for example slicing.

Reply:
Ah, so #305?

Reply:
That issue is about Catalyst gradients, I think Romain is talking about JAX gradients (which are run in the frontend on the jaxpr, hence the primitives need gradient rules).

Reply:
Oh, got it!

Reply:
Yes exactly what David said, qjit with jax.grad and slicing

Reply:
```
qjit(jax.grad(f))(x)
```

Reply:
Just checking if we have a resolution on this particular comment thread :)

Reply:
@grwlf How is the upstream PR looking for this patch? 
In the meantime, can we just attach the original gradient rule to the patched primitive?